src/runtime/refiner.ts

```diff
@@ -64,6 +68,56 @@ export class Refinement {
     }
   }
 
+  static unionOf(ref1: Refinement, ref2: Refinement): Refinement {
+    const expr1 = ref1 && ref1.expression;
+    const expr2 = ref2 && ref2.expression;
+    let refinement = null;
+    if (expr1 && expr2) {
+      const bothExpr = new BinaryExpression(expr1, expr2, new RefinementOperator(Op.OR));
+      refinement = new Refinement(bothExpr);
+    }
+    return refinement;
+  }
+
+  static intersectionOf(ref1: Refinement, ref2: Refinement): Refinement {
+    const expr1 = ref1 && ref1.expression;
+    const expr2 = ref2 && ref2.expression;
+    let refinement = null;
+    if (expr1 && expr2) {
+      const bothExpr = new BinaryExpression(expr1, expr2, new RefinementOperator(Op.AND));
+      refinement = new Refinement(bothExpr);
+    } else if (expr1 || expr2) {
+      refinement = new Refinement(expr1 || expr2);
+    }
+    return refinement;
+  }
+
+  containsField(fieldName: string): boolean {
+    return this.expression.containsField(fieldName);
+  }
+
+  fieldNameIfUnivariate(): string {
+    return this.expression.fieldNameIfUnivariate();
+  }
+
+  // checks if a is more specific than b, returns null if can't be determined
+  static isMoreSpecific(a: Refinement, b: Refinement): boolean {
```

Comment:
I was debating this for a while. The thing is 'a' can be null (in which case we can't call the member function) and I thought it would be cleaner to have one function that handles all cases rather than handling the case where 'a' is null separately while still handling the case where 'b' is null inside the member function.

Reply:
Ah right, that makes sense. Static is fine then.